ddtrace/internal/telemetry/telemetry_writer.py

```diff
@@ -0,0 +1,234 @@
+            "DD-Telemetry-API-Version": "v1",
+        }
+
+    def _send_request(self, request):
+        # type: (Dict) -> httplib.HTTPResponse
+        """
+        Sends a telemetry request to the trace agent
+        """
+        conn = get_connection(self.AGENT_URL)
+        rb_json = self.encoder.encode(request)
+        resp = None
+        with StopWatch() as sw:
+            try:
+                conn.request("POST", self.ENDPOINT, rb_json, self._get_headers(request["request_type"]))
+                resp = get_connection_response(conn)
+
+                t = sw.elapsed()
+                if t >= self.interval:
+                    log_level = logging.WARNING
+                else:
+                    log_level = logging.DEBUG
+
+                log.log(
+                    log_level,
+                    "sent %d in %.5fs to %s/%s. response: %s",
+                    len(rb_json),
+                    t,
+                    self.AGENT_URL,
+                    self.ENDPOINT,
+                    resp.status,
+                )
+            finally:
+                conn.close()
+        return resp
+
+    def flush_integrations_queue(self):
+        # type () -> List[Dict]
+        """Returns a list of all integrations queued by integration_event"""
+        with self._lock:
+            integrations = self._integrations_queue
+            self._integrations_queue = []
+        return integrations
+
+    def flush_events_queue(self):
+        # type () -> List[TelemetryRequest]
+        """Returns a list of all integrations queued by classmethods"""
+        with self._lock:
+            requests = self._events_queue
+            self._events_queue = []
+        return requests
+
+    def queued_events(self):
+        # type () -> List[Dict]
+        return self._events_queue
+
+    def queued_integrations(self):
+        # type () -> List[Dict]
+        return self._integrations_queue
+
+    def periodic(self):
+        telemetry_requests = self.flush_events_queue()
+
+        integrations = self.flush_integrations_queue()
+        if integrations:
+            integrations_request = TelemetryRequest(app_integrations_changed_telemetry_request(integrations), 0)
+            telemetry_requests.append(integrations_request)
+
+        requests_failed = []  # type: List[Dict]
+        for telemetry_request in telemetry_requests:
+            request, fail_count = telemetry_request
+
+            resp = self._send_request(request)
+            if resp.status >= 300 and fail_count + 1 < self.MAX_FAIL_COUNT:
+                requests_failed.append(TelemetryRequest(request, fail_count + 1))
+
+        with self._lock:
+            # add failed requests back to the requests queue
+            self._events_queue.extend(requests_failed)
+
+    def shutdown(self):
+        # type: () -> None
+        self.app_closed_event()
+        self.periodic()
+
+    @classmethod
+    def add_event(cls, request):
```

Comment:
TODO: have this take in `payload, payload_type`, and then inside of here use `_create_telemetry_request(payload, payload_type, seq_id)` to create and enqueue the event request.